### ISSUE TITLE
Tailwind class prefix support

### ISSUE THREAD

**Issue by zakstucke** [gaucho-labs/tailwind-fuse#12]
Cool library! Can prefixes be supported?

I use `"tw-"` prefix on tailwind classes. 
`tw_merge!()` seems to not recognise prefixed classes correctly, and acts like `tw_join!()` instead.

E.g. `tw_merge!("tw-bg-red-500", "tw-bg-black-500")` comes out as `"tw-bg-red-500 tw-bg-black-500"`

**Comment by nicoburniske** [comment]
I have to improve the docs for this. 

The macro `tw_merge!` does not support prefix, but the inner function does. 

Take a look at this https://docs.rs/tailwind_fuse/latest/tailwind_fuse/merge/fn.tw_merge_with_options.html

**Comment by nicoburniske** [comment]
I am not sure how I could support this with the macro. If you have any ideas lmk

**Comment by nicoburniske** [comment]
what do you think about this? It's pretty annoying to supply the options everywhere though. The arrow symbol can be replaced for something else if there's a better alternative

```rust
#[test]
fn test_override_config_macro() {
    let config = MergeOptions {
        prefix: "tw-",
        separator: "|",
    };

    let result = tw_merge!(config => "hover|lg|tw-bg-blue-100", "hover|lg|tw-bg-red-500");

    assert_eq!("hover|lg|tw-bg-red-500", result);
}

```

**Comment by zakstucke** [comment]
Hi @nicoburniske, sorry for only looping back now.

I'm not a big fan of your `config => ` experiment, it think it would at a *lot* of bloat and mental overhead around a codebase.

Here's my full solution I'm currently using:

```rust
// prelude.rs

macro_rules! tw_mymerge {
    ($($item:expr),+ $(,)?) => {{
        let joined = tailwind_fuse::tw_join!($($item),+);
        tailwind_fuse::merge::tw_merge_with_options(joined.as_str(), tailwind_fuse::merge::MergeOptions {
            prefix: "tw-",
            separator: ":",
        })
    }};
}
```

```toml
# clippy.toml

disallowed-methods = ["tailwind_fuse::IntoTailwindClass::with_class"]
disallowed-macros = ["tailwind_fuse::tw_merge"]
```

Effectively I'm just banning usage of `tw_merge` and `with_class`, to make sure I'm only using my `tw_mymerge!()` version that has the config built into it. I'm not seeing a better way than this right now, but this works great in the sense I don't have to worry about it again!

Feel free to close this if I've hit on the least painful way to deal with this :)

**Comment by zakstucke** [comment]
Awesome! I'll give that a go, means for me with my macro and the clippy lint on tw_merge this is seems to be all sorted.

> You're spot on with the tw_mymerge impl, though I am not sure it's a good idea to assume that users can write their own macros 😄

Haha makes sense. To be honest I've never seen a different prefix that `tw-`, this is probably the most common case by far, could maybe make a `tw-prefix` feature which uses my macro instead of current when feature is enabled? This could be done for the most common prefixes people use (feature for each), and anyone else recommended to use their own macro. Not perfect but would probs cover most cases.

**Comment by zakstucke** [comment]
@nicoburniske I actually just thought of a cool way to deal with this:

```rust
use linkme::distributed_slice;

// UPSTREAM IN TAILWIND_FUSE
#[distributed_slice]
pub static PREFIX_OVERRIDE: [&'static str];

// UPSTREAM IN TAILWIND_FUSE
// Internally in tailwind fuse use this to get prefix, either the one from the slice a user set, or defaulting to nothing like normal.
pub fn get_prefix() -> &'static str {
    PREFIX_OVERRIDE.first().copied().unwrap_or("")
}

/// UPSTREAM IN TAILWIND_FUSE
/// Hiding the implementation behind a simple macro.
macro_rules! set_tw_custom_prefix {
    ($prefix:expr) => {
        const _: () = {
            use linkme::distributed_slice;
            #[distributed_slice(PREFIX_OVERRIDE)]
            static TAILWIND_FUSE_DEFAULT_PREFIX: &str = $prefix;
        };
    };
}

// DOWNSTREAM IN USER CODE
set_tw_custom_prefix!("tw-");
```

https://github.com/dtolnay/linkme

It's pretty cool this crate and I use this approach for something else too. Could it work here and remove the need for any of this?

**Comment by nicoburniske** [comment]
Wow that's insanely cool @zakstucke 

This seems like a better approach to me than the env variables PR I linked because it's code first.

**Comment by zakstucke** [comment]
Yeah I saw that come in at the same time I sent this! I'd agree imo this is a slightly cleaner approach, but I think either would work.

**Comment by nicoburniske** [comment]
Hm @zakstucke linkme doesn't have WASM support unfortunately. 

https://github.com/dtolnay/linkme/issues/6

**Comment by zakstucke** [comment]
Ah such a shame, I hadn't seen that.
Well, your approach still sounds like a good way forward!

**Comment by nicoburniske** [comment]
@zakstucke see https://github.com/gaucho-labs/tailwind-fuse/pull/21

**Comment by nicoburniske** [comment]
Both your issues (Option in macros, and Prefix) should be solved in `0.3.0`

**Comment by zakstucke** [comment]
@nicoburniske OnceLock was a great middle ground! 

I've just setup with 0.3 and both this prefixing and the re-support of `Option` work seamlessly, thanks so much for your hard work! 🎉
